java/org/apache/catalina/webresources/AbstractSingleArchiveResourceSet.java

```diff
@@ -63,29 +63,31 @@ public AbstractSingleArchiveResourceSet(WebResourceRoot root, String webAppMount
 
     @Override
     protected HashMap<String,JarEntry> getArchiveEntries(boolean single) {
-        synchronized (archiveLock) {
-            if (archiveEntries == null && !single) {
-                JarFile jarFile = null;
-                archiveEntries = new HashMap<>();
-                try {
-                    jarFile = openJarFile();
-                    Enumeration<JarEntry> entries = jarFile.entries();
-                    while (entries.hasMoreElements()) {
-                        JarEntry entry = entries.nextElement();
-                        archiveEntries.put(entry.getName(), entry);
-                    }
-                } catch (IOException ioe) {
-                    // Should never happen
-                    archiveEntries = null;
-                    throw new IllegalStateException(ioe);
-                } finally {
-                    if (jarFile != null) {
-                        closeJarFile();
+        if (archiveEntries == null && !single) {
+            synchronized (archiveLock) {
+                if (archiveEntries == null) {
```

Comment:
`archiveEntries` in `AbstractArchiveResourceSet.archiveEntries` is not `volatile` so this could still be read as `null` inside lock by a reading thread just after another thread writes this.  Should we mark it `volatile`?

https://wiki.sei.cmu.edu/confluence/display/java/LCK10-J.+Use+a+correct+form+of+the+double-checked+locking+idiom

Adding volatile should not slow down the reads much, only the one-time assignment:

https://stackoverflow.com/questions/4633866/is-volatile-expensive

Reply:
The consequence would be a possible double-initialization?  I think that risk existed before?  I'd be game to make that adjustment, but I'm thinking the this change should just be to avoid the synchronization block on every get. Since my intent is to submit this to apache, I want to keep the change as focused as I can.

Reply:
Hmm... good point that this existed before.  Ok, sounds good.  I was thinking of some weird error edge cases where you get a double-initialization, different threads store these equivalent but not-reference-equal `JarEntry`s, and then someone tries to compare `JarEntry`s across threads by reference, which fails, leading to some subtle bugs.  Kind of similar to how we have bugs occasionally when comparing classes across class loaders.

Reply:
@ghollies ☝️ it's possible the issue Kevin is talking about here is something we need to look at after all?  It'd be an easy thing to test if we can recreate the problems we're seeing on home and elsewhere.

Reply:
@ghollies @fluetm it would be possible for this method to return an incomplete Map of `archiveEntries`. If `archiveEntries` has been initialized on line 70 but has not yet had all the entries added and another thread calls this method, it would skip the synchronized block entirely and return the incomplete Map. I am not sure if that would cause the problems though, just something I noticed

Reply:
Yeah, @jack-nolan is right.  Better to build the map in a temp variable and then assign `archiveEntries` at the very end.

Reply:
Ugh...yeah that's a good catch @jack-nolan, I missed that 🤷‍♂  They were getting away with it because the synchronized block was so broad.  It would certainly fit the profile of the issues that were reported...occasional race conditions that led to classes or resource files not being found.  Should be easy enough to correct.